pkg/controller/imagelistpulljob/imagelistpulljob_controller.go

```diff
@@ -0,0 +1,425 @@
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("update ImageListPullJob status error: %v", err)
+		}
+		resourceVersionExpectations.Expect(job)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileImageListPullJob) computeImagePullJobActions(job *appsv1alpha1.ImageListPullJob, imagePullJobs map[string]*appsv1alpha1.ImagePullJob) ([]*appsv1alpha1.ImagePullJob, []*appsv1alpha1.ImagePullJob) {
+	var needToDelete, needToCreate []*appsv1alpha1.ImagePullJob
+	//1. need to create
+	images, needToDelete := r.filterImagesAndImagePullJobs(job, imagePullJobs)
+	needToCreate = r.newImagePullJobs(job, images)
+	// some images delete from ImageListPullJob.Spec.Images
+	for image, imagePullJob := range imagePullJobs {
+		if !slice.ContainsString(job.Spec.Images, image, nil) {
+			needToDelete = append(needToDelete, imagePullJob)
+			key := types.NamespacedName{Namespace: job.Namespace, Name: job.Name}.String()
+			scaleExpectations.ExpectScale(key, expectations.Delete, imagePullJob.Spec.Image)
+		}
+	}
+	return needToCreate, needToDelete
+}
+
+func (r *ReconcileImageListPullJob) calculateStatus(job *appsv1alpha1.ImageListPullJob, imagePullJobs map[string]*appsv1alpha1.ImagePullJob) *appsv1alpha1.ImageListPullJobStatus {
+	var active, completed, succeeded int32
+	// record the failed image pull job status
+	var failedImageStatuses []*appsv1alpha1.FailedImageStatus
+
+	for _, imagePullJob := range imagePullJobs {
+		if imagePullJob.Status.StartTime == nil {
+			continue
+		}
+
+		if imagePullJob.Status.Active > 0 {
+			active = active + 1
+		}
+
+		if imagePullJob.Status.Desired == (imagePullJob.Status.Failed + imagePullJob.Status.Succeeded) {
+			completed = completed + 1
+			//
+			if imagePullJob.Status.Failed > 0 {
+				failedImagePullJobStatus := &appsv1alpha1.FailedImageStatus{
+					ImagePullJob: imagePullJob.Name,
+					Name:         imagePullJob.Spec.Image,
+					Message:      fmt.Sprintf("Please check for details which nodes failed by using %s.", imagePullJob.Name),
+				}
+				failedImageStatuses = append(failedImageStatuses, failedImagePullJobStatus)
+			}
+			resourceVersionExpectations.Delete(imagePullJob)
+		}
+
+		if imagePullJob.Status.Desired == imagePullJob.Status.Succeeded {
+			succeeded = succeeded + 1
+			resourceVersionExpectations.Delete(imagePullJob)
+		}
+	}
+
+	// 4. status
+	newStatus := &appsv1alpha1.ImageListPullJobStatus{
+		Desired:             int32(len(job.Spec.Images)),
+		Active:              active,
+		Completed:           completed,
+		Succeeded:           succeeded,
+		Status:              fmt.Sprintf("%d/%d", succeeded, completed),
+		StartTime:           job.Status.StartTime,
+		FailedImageStatuses: failedImageStatuses,
+	}
+
+	now := metav1.NewTime(r.clock.Now())
+	if newStatus.StartTime == nil {
+		newStatus.StartTime = &now
+	}
+	// the same behavior as image pull job
+	if job.Spec.CompletionPolicy.Type != appsv1alpha1.Never && (newStatus.Desired-newStatus.Completed) == 0 {
```

Comment:
consider simplify the code block:
newStatus.Desired-newStatus.Completed) == 0 

to more simple form
newStatus.Desired == newStatus.Completed

Reply:
all done